enhancements/kata-containers/kata-containers-dev-preview.md

```diff
@@ -0,0 +1,276 @@
+
+**So What do we want to do with this?**
+
+We start by creating a container image with only the kata upstream components as static binaries (no RHEL-AV components). The kata operator will pull down this image and mount it in the host mount namespace, i.e. without running it as a container. We then use the RHCOS extensions for deploying the qemu-kiwi RHEL-AV RPMs.
+
+**Pros:**
+1. RHEL-AV depends on libraries and other tools which are all available as RPMs. Turning this into a static image instead of RPMs creates a huge single binary which we avoid with extensions
+2. RHEL-AV has been tested / validated as RPMs and not a single binary which means we saves the extra overhead
+3. RHEL-AV RPMs can be installed only by the customers who would be interested in using the kata runtime. QEMU (and its dependencies) get updated as part of the machine-os-content updates
+4. There's no extra work related to properly loading qemu-kiwi dependencies, as all the packages would be installed in the same PATH its build in
+5. Better control of the versions that are installed both for qemu (specific version for a given host) and for the dependencies (tracked by rpm)
+
+**Cons:**
+1. ~20 mb extra, installed in the host, for those who'd be using kata runtime
+2. machine-os-content would still carry this until there is a separate machine-os-content- extensions container
+
+
+
+### KataContainers Operator development
+The goal is to develop a kata operator that can be used to manage the entire lifecycle of kata components in an OpenShift cluster. We will have a controller that watches for a kata **custom resource (CR)** and a daemon-set that acts upon changes to the (CR) and do the required work on the worker nodes (install, uninstall update,...).
+
+Via the CR it will be possible to select a subset of worker nodes.
+
+For deploying binaries on the host the current idea is to use a container image that can be mounted on the host level.  
+
+The goal is to be able to install the operator via OperatorHub.
+
+The Operator will:
+- Create a crio drop-in config file via machine config objects;
+- Automatically select the payload image that contains correct version of the kata binaries for the given version of the OpenShift;
+- Configure a `RuntimeClass` called `kata` which can be used to deploy workload that uses kata runtime;
+- Create a machine-config object to enable the `qemu-kiwi` RHCOS extension.
+
+The `RuntimeClass` and payload image names will be visible in the CR status.
+
+#### KataContainers Operator Goals
+- Create an API which supports installation of Kata Runtime on all or selected worker nodes
+- Configure CRI-O to use Kata Runtime on those worker nodes
+- Installation of the runtimeClass on the cluster, as well as of the required components for the runtime to be controlled by the orchestration layer.
+- Updates the Kata runtime
+- Uninstall Kata Runtime and reconfigure CRI-O to not use it.
```

Comment:
If done via extensions, this would all be handled by MCO reconciliation.   See also https://blog.verbum.org/2020/08/22/immutable-%E2%86%92-reprovisionable-anti-hysteresis/

I think more broadly, if everything goes via MCO functionality then e.g. node changes will be fully coordinated.  Having more host daemonsets etc. can create synchronization issues.

Reply:
When you say "everything goes via MCO functionality" you mean ... just forget the operator and let teach the MCO to deal with Kata Containers' needs, entirely?

While I think that's a really good idea, I don't think it's do-able for 4.8 timeframe.

Reply:
Ah no, I just mean that the operator installs machineconfigs, rather than reaching out to each node individually.  This is a pattern followed by for example the cluster-node-tuning-operator.

Reply:
Oh, that we do. :-)